bazaar/rpc/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Encointer.  If not, see <http://www.gnu.org/licenses/>.
+
+use jsonrpc_core::{Error, ErrorCode, Result};
+use jsonrpc_derive::rpc;
+use sp_api::{Decode, Encode, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+use std::sync::Arc;
+use sc_rpc::DenyUnsafe;
+
+use encointer_bazaar_rpc_runtime_api::BazaarApi as BazaarRuntimeApi;
+use encointer_primitives::{communities::CommunityIdentifier, bazaar::BusinessIdentifier};
+use encointer_primitives::bazaar::{OfferingData, BusinessData};
+
+#[rpc]
+pub trait BazaarApi<BlockHash, AccountId>
+    where AccountId: 'static + Encode + Decode + Send + Sync, {
+    #[rpc(name = "bazaar_getBusinesses")]
+    fn get_businesses(&self, cid: CommunityIdentifier, at: Option<BlockHash>) -> Result<Vec<BusinessData>>;
+    #[rpc(name = "bazaar_getOfferings")]
+    fn get_offerings(&self, cid: CommunityIdentifier, at: Option<BlockHash>) -> Result<Vec<OfferingData>>;
+    #[rpc(name = "bazaar_getOfferingsForBusiness")]
+    fn get_offerings_for_business(&self, bid: BusinessIdentifier<AccountId>, at: Option<BlockHash>) -> Result<Vec<OfferingData>>;
+}
+
+pub struct Bazaar<Client, Block, AccountId> {
+    client: Arc<Client>,
+    _marker: std::marker::PhantomData<(Block, AccountId)>,
+    deny_unsafe: DenyUnsafe,
+}
+
+impl<Client, Block, AccountId> Bazaar<Client, Block, AccountId>
+    where
+{
+    /// Create new `Bazaar` instance with the given reference to the client.
+    pub fn new(client: Arc<Client>) -> Self {
+        Bazaar {
+            client,
+            _marker: Default::default(),
+            deny_unsafe: DenyUnsafe::Yes
+        }
+    }
+}
+
+impl<Client, Block, AccountId> BazaarApi<<Block as BlockT>::Hash, AccountId> for Bazaar<Client, Block, AccountId>
+    where
+        AccountId: 'static + Encode + Decode + Send + Sync + Copy,
+        Block: BlockT,
+        Client: Send + Sync + 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+        Client::Api: BazaarRuntimeApi<Block, AccountId> {
+    fn get_businesses(&self, cid: CommunityIdentifier, at: Option<<Block as BlockT>::Hash>) -> Result<Vec<BusinessData>> {
+        if let Err(err) = self.deny_unsafe.check_if_safe() {
+            return Err(Error::from(err));
+        }
+
+        let api = self.client.runtime_api();
+        let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
+        return Ok(api.get_businesses(&at, &cid)
+            .map_err(runtime_error_into_rpc_err)?.iter()
+            .map(|bid| bid.1.clone())
+            .collect());
+    }
+
+    fn get_offerings(&self, cid: CommunityIdentifier, at: Option<<Block as BlockT>::Hash>) -> Result<Vec<OfferingData>> {
```

Comment:
`check_if_safe()?;` (note the question mark) implicitly performs the `Error::from(error)` conversion.